### PR TITLE
feat(vision): detect+classify combo, session TTL cleanup, COCO dataset support

### PR DESCRIPTION
**Pull request by rgthelen**
Three vision API improvements from real usage patterns:

## 1. Detect + Classify Combo Endpoint
`POST /v1/vision/detect_classify` — YOLO detect → auto-crop each bbox → CLIP classify per crop → unified results. Eliminates 3+ round-trips for the most common multi-model pattern.

```json
{
  "image": "<base64>",
  "detection_model": "yolov8n",
  "classification_model": "clip-vit-base",
  "classes": ["robin", "eagle", "sparrow"],
  "confidence_threshold": 0.5
}
```

Returns per-detection classification with both detection and classification confidence, bounding boxes, and top-k scores.

## 2. Streaming Session TTL Cleanup
- Sessions auto-expire after 60s of no frames (configurable `SESSION_TTL_SECONDS`)
- Background cleanup task every 15s
- `GET /v1/vision/stream/sessions` — list active sessions with idle time, frames processed, chain config

## 3. COCO JSON Dataset Support
Training endpoint auto-detects COCO JSON format and converts to YOLO internally:
- COCO bbox `[x, y, w, h]` (absolute) → YOLO `[class, x_center, y_center, w, h]` (normalized)
- Auto train/val split (80/20 default)
- Generates `data.yaml` for ultralytics
- Supports public datasets (LVIS, iNaturalist, OpenImages) directly

## Files Changed (8 files, +498 lines)
- **New**: `routers/vision/detect_classify.py` — combo endpoint
- **New**: `vision_training/coco_converter.py` — COCO→YOLO conversion
- **Modified**: `routers/vision/streaming.py` — TTL cleanup + sessions list
- **Modified**: `vision_training/trainer.py` — auto-detect COCO format
- **Modified**: `routers/vision/__init__.py`, `server.py` — wiring
- **Modified**: `server/api/routers/vision/pipeline.py`, `pipeline_service.py` — server proxy

### CODE REVIEW

File: runtimes/universal/routers/vision/detect_classify.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
+"""Detect+Classify combo endpoint — YOLO detect → crop → CLIP classify per crop."""
+
+import io
+import logging
+import time
+from collections.abc import Callable, Coroutine
+from typing import Any
+
+import numpy as np
```
[the remark this answers is not in the document]
Comment (rgthelen) [reply runtimes/universal/routers/vision/detect_classify.py]: The `np` import was already removed in a previous commit.

File: runtimes/universal/routers/vision/streaming.py [hunk 2]
```diff
@@ -63,9 +64,38 @@ class StreamSession:
...
+                    f"(idle {now - session.last_frame_at:.0f}s, "
+                    f"{session.frames_processed} frames processed)"
+                )
+
+
+def start_session_cleanup() -> None:
+    """Start the background session cleanup task. Call once at server startup."""
+    global _cleanup_task
+    if _cleanup_task is None or _cleanup_task.done():
+        _cleanup_task = asyncio.create_task(_session_cleanup_loop())
```
[the remark this answers is not in the document]
Comment (rgthelen) [reply runtimes/universal/routers/vision/streaming.py:98]: `_cleanup_task` stores the `asyncio.Task` reference to prevent garbage collection of the background cleanup coroutine. Without it, the task can be GC'd before completion. This is the standard pattern per [Python docs](https://docs.python.org/3/library/asyncio-task.html#creating-tasks).

File: runtimes/universal/routers/vision/detect_classify.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+from fastapi import APIRouter, HTTPException
+from PIL import Image
+from pydantic import BaseModel, Field
+
+from services.error_handler import handle_endpoint_errors
+
+from .utils import decode_base64_image
+
+logger = logging.getLogger(__name__)
+router = APIRouter(tags=["vision-detect-classify"])
```
[the remark this answers is not in the document]
Comment (rgthelen) [reply runtimes/universal/routers/vision/detect_classify.py]: `np` was already removed. Keeping `logger` — it's the standard module-level logger pattern and will be used as the module matures.

File: server/api/routers/vision/pipeline.py [hunk 4]
```diff
@@ -10,10 +10,44 @@
...
+    det_cls = [c.strip() for c in detection_classes.split(",") if c.strip()] if detection_classes else None
+    return await VisionPipelineService.detect_classify({
+        "image": image,
+        "detection_model": detection_model,
+        "classification_model": classification_model,
+        "classes": cls_list,
+        "confidence_threshold": confidence_threshold,
+        "detection_classes": det_cls,
+        "top_k": top_k,
+    })
```
[the remark this answers is not in the document]
Comment (rgthelen) [reply server/api/routers/vision/pipeline.py:44]: Added model name validation in 152283eb — both `detection_model` and `classification_model` are now validated as non-empty, alongside the existing `classes` validation.

File: runtimes/universal/routers/vision/detect_classify.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        # Skip tiny crops
+        if (x2 - x1) < request.min_crop_px or (y2 - y1) < request.min_crop_px:
+            continue
+
+        crop = pil_image.crop((x1, y1, x2, y2))
+
+        # Convert crop to bytes for the classifier
+        buf = io.BytesIO()
+        crop.save(buf, format="JPEG", quality=90)
+        crop_bytes = buf.getvalue()
```
[the remark this answers is not in the document]
Comment (rgthelen) [reply runtimes/universal/routers/vision/detect_classify.py:142]: Fixed in d464668f — crops are now converted to RGB before JPEG encoding.

File: runtimes/universal/routers/vision/detect_classify.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    return DetectClassifyResponse(
+        results=results,
+        total_detections=total_detections,
+        classified_count=len(results),
+        detection_model=request.detection_model,
+        classification_model=request.classification_model,
+        detection_time_ms=det_time,
+        classification_time_ms=cls_time,
+        total_time_ms=(time.perf_counter() - total_start) * 1000,
+    )
```
[the remark this answers is not in the document]
Comment (rgthelen) [reply runtimes/universal/routers/vision/detect_classify.py:171]: Integration tests run against the live server — 13/13 passing through the full proxy chain (server → runtime detect_classify endpoint).

File: runtimes/universal/server.py [hunk 7]
```diff
@@ -1192,9 +1194,11 @@ async def load_tts(
 set_file_image_getter(get_file_images)
 set_detection_loader(load_detection_model)
 set_classification_loader(load_classification_model)
+set_detect_classify_loaders(load_detection_model, load_classification_model)
 set_streaming_detection_loader(load_detection_model)
 set_vision_models_dir(VISION_MODELS_DIR)
 set_model_export_loader(load_detection_model)
+start_session_cleanup()
```
[the remark this answers is not in the document]
Comment (rgthelen) [reply runtimes/universal/server.py]: Fixed in cdbabcfa — `start_session_cleanup()` moved into the `lifespan()` context manager.

File: runtimes/universal/vision_training/coco_converter.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+        # Copy image if it exists
+        src_path = images_dir / filename
+        if src_path.exists():
+            shutil.copy2(str(src_path), str(img_dst / filename))
+        else:
+            # Try without subdirectory
+            src_path = images_dir / Path(filename).name
+            if src_path.exists():
+                shutil.copy2(str(src_path), str(img_dst / Path(filename).name))
+            else:
```
[the remark this answers is not in the document]
Comment (rgthelen) [reply runtimes/universal/vision_training/coco_converter.py:126]: Fixed in d464668f — `file_name` is now sanitized with `os.path.basename()` and validated to stay within the output directory.

File: runtimes/universal/vision_training/coco_converter.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+                logger.warning(f"Image not found: {filename}")
+                skipped += 1
+                continue
+
+        # Convert annotations to YOLO format
+        annotations = annots_by_image.get(img_id, [])
+        label_name = Path(filename).stem + ".txt"
+        lines: list[str] = []
```
[the remark this answers is not in the document]
Comment (rgthelen) [reply runtimes/universal/vision_training/coco_converter.py]: Valid point. The current `basename()` stripping is intentional for security (path traversal prevention), but it could lose subdirectory structure info. Noted as a future improvement — we could preserve relative subdirs within a safe root.

File: runtimes/universal/routers/vision/streaming.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rgthelen) [reply runtimes/universal/routers/vision/streaming.py:98]: Fixed in cdbabcfa — `start_session_cleanup()` moved into `lifespan()`.

File: runtimes/universal/vision_training/coco_converter.py [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+
+        # Determine split
+        is_train = img_id in train_ids
+        img_dst = train_images if is_train else val_images
+        lbl_dst = train_labels if is_train else val_labels
+
+        # Copy image if it exists
+        src_path = images_dir / filename
+        if src_path.exists():
+            shutil.copy2(str(src_path), str(img_dst / filename))
```
[the remark this answers is not in the document]
Comment (rgthelen) [reply runtimes/universal/vision_training/coco_converter.py]: Fixed in d464668f — `file_name` is sanitized via `os.path.basename()` with output directory containment check.

File: server/api/routers/vision/pipeline.py [hunk 11]
```diff
@@ -10,10 +10,44 @@
...
+    image: str = Form(..., description="Base64-encoded image"),
+    detection_model: str = Form(default="yolov8n"),
+    classification_model: str = Form(default="clip-vit-base"),
+    classes: str = Form(..., description="Comma-separated classification classes"),
+    confidence_threshold: float = Form(default=0.5),
+    detection_classes: str | None = Form(default=None),
+    top_k: int = Form(default=3),
+) -> dict[str, Any]:
+    """Detect objects then classify each crop — single round-trip."""
+    cls_list = [c.strip() for c in classes.split(",") if c.strip()]
```
[the remark this answers is not in the document]
Comment (rgthelen) [reply server/api/routers/vision/pipeline.py:32]: Classes validation was already present. Added model name validation in 152283eb.

File: runtimes/universal/vision_training/coco_converter.py [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+            continue
+
+        # Determine split
+        is_train = img_id in train_ids
+        img_dst = train_images if is_train else val_images
+        lbl_dst = train_labels if is_train else val_labels
+
+        # Copy image if it exists — try original relative path first, then basename
+        src_path = images_dir / raw_filename
+        dst_name = filename  # Always use sanitized basename for output
```
[the remark this answers is not in the document]
Comment (rgthelen) [reply runtimes/universal/vision_training/coco_converter.py:119]: Valid edge case. The basename flattening is primarily for path traversal prevention. Collision detection for same-named files from different subdirs is a reasonable future improvement — low risk for typical COCO datasets but worth addressing.

**Comment by mhamann** [comment]
🚨 **Critical Security Issue - Path Traversal Vulnerability**

**File:** `coco_converter.py` line ~108

The COCO converter uses `file_name` directly from the dataset for file paths without proper validation. This allows potential path traversal attacks where malicious COCO datasets could read/write files outside the intended directory.

**Current code:**
```python  
src_path = images_dir / raw_filename
```

**Recommended fix:**
```python
src_path = images_dir / raw_filename  
if not src_path.resolve().is_relative_to(images_dir.resolve()):
    logger.warning(f"Skipping file outside images directory: {raw_filename}")
    continue
```

This vulnerability could allow attackers to access sensitive files or corrupt data outside the images directory.

**Comment by mhamann** [comment]
🚨 **Critical Reliability Issue - Server Startup Failure**

**File:** `streaming.py` line ~98

The code attempts to create an asyncio task at module import time before an event loop exists:

```python
_cleanup_task = asyncio.create_task(_session_cleanup_loop())
```

**Impact:** This will cause `RuntimeError: no running event loop` and prevent the server from starting entirely.

**Recommended fix:**
```python
def start_session_cleanup() -> None:
    global _cleanup_task
    if _cleanup_task is None or _cleanup_task.done():
        try:
            _cleanup_task = asyncio.create_task(_session_cleanup_loop())
        except RuntimeError:
            logger.warning("No event loop running, session cleanup not started")
```

The current approach in `server.py` calling this in the lifespan hook is correct, but the function needs proper event loop handling.

**Comment by mhamann** [comment]
⚠️ **High Priority Issues**

## File Collision Risk
**File:** `coco_converter.py` line ~108

Flattening all image paths to basenames can cause silent file overwrites when different subdirectories contain identically named files.

**Current:** `dst_name = Path(raw_filename).name`

**Risk:** Files like `images/train/cat.jpg` and `images/val/cat.jpg` both become `cat.jpg`, causing data loss.

**Fix:** Either preserve directory structure safely or detect collisions:
```python
if (img_dst / dst_name).exists():
    logger.error(f"File collision detected: {dst_name}")
    # Handle collision (skip, rename, or preserve structure)
```

## Missing Test Coverage
This PR adds 518 lines of new code with zero tests. For production code with security implications, comprehensive test coverage is essential:

- Unit tests for COCO converter with malicious path scenarios
- Integration tests for detect+classify endpoint  
- Session cleanup timing tests
- Error handling validation

**Recommendation:** Add test coverage before merging to prevent regressions and ensure security fixes work correctly.

**Review by rgthelen — COMMENTED** [review]
@mhamann 

## Addressing Review Feedback

Thanks for the thorough review Matt. All 4 issues addressed:

### 1. ✅ Server Startup Blocker (streaming.py)
Fixed in `cdbabcfa` — `start_session_cleanup()` moved into `lifespan()` async context manager.

### 2. ✅ Path Traversal (coco_converter.py)
Fixed in `d464668f` — `file_name` sanitized via `os.path.basename()` + reject `..`, `:`, `\` patterns + `is_relative_to()` containment check on resolved paths.

### 3. ✅ File Collision Risk (coco_converter.py)
Fixed in `17d2016b` — Added basename collision detection. When different subdirectories contain same-named files, the second is skipped with a warning instead of silently overwriting.

### 4. ✅ Test Coverage
Added in `21867d65` — **39 unit tests** across 3 new test files:
- **`test_coco_converter.py`** (18 tests): path traversal rejection, bbox normalization, split ratio, basename collision detection, validation, edge cases
- **`test_detect_classify.py`** (8 tests): RGBA handling, tiny crop skip, empty classes, no loaders, full detect→classify flow
- **`test_streaming_vision.py`** (13 tests): session lifecycle, max sessions, TTL cleanup, cascade config, cooldown suppression, frame counting

All 39/39 passing.
